File: docx-parser-converter-js/src/parsers/tables/table-cell-properties-parser.ts
/**
 * Table cell properties parser for DOCX documents
 * Parses table cell properties from w:tcPr elements
 */

import type { 
  TableCellProperties, 
  TableWidth, 
  TableCellBorders, 
  ShadingProperties, 
  MarginProperties 
} from '@/models/table-models.js';
import { 
  TableCellPropertiesModel, 
  TableWidthModel, 
  TableCellBordersModel, 
  ShadingPropertiesModel, 
  MarginPropertiesModel,
  BorderPropertiesModel
} from '@/models/table-models.js';
import { convertTwipsToPoints } from '@/utils/unit-conversion.js';
import { getFirstChildElement } from '@/utils/xml-utils.js';
import { BaseParser } from '../base-parser.js';
import { extractBooleanProperty } from '../helpers/common-helpers.js';

/**
 * Table cell properties parser class
 */
export class TableCellPropertiesParser extends BaseParser<TableCellProperties> {
  constructor(options: Record<string, unknown> = {}) {
    super('TableCellPropertiesParser', options);
  }

  /**
   * Parse XML object into TableCellProperties model
   * @param xmlObj - Parsed XML object containing w:tcPr element
   * @returns Promise resolving to TableCellProperties model
   */
  protected async parseInternal(xmlObj: Record<string, unknown>): Promise<TableCellProperties> {
    // Extract w:tcPr element
    let tcPrElement: Record<string, unknown>;

    if (xmlObj['w:tcPr']) {
      const tcPrValue = xmlObj['w:tcPr'];
      
      // Handle case where w:tcPr is an array
      if (Array.isArray(tcPrValue)) {
        if (tcPrValue.length === 0) {
          throw new Error('Empty w:tcPr array found in XML');
        }
        tcPrElement = tcPrValue[0] as Record<string, unknown>;
      } else {
        tcPrElement = tcPrValue as Record<string, unknown>;
      }
    } else if (xmlObj && typeof xmlObj === 'object' && Object.keys(xmlObj).length > 0) {
      // If the root object itself is the w:tcPr element
      tcPrElement = xmlObj;
    } else {
      throw new Error('No w:tcPr element found in XML');
    }

    return this.parseTableCellPropertiesElement(tcPrElement);
  }

  /**
   * Parse table cell properties element
   * @param tcPrElement - w:tcPr element
   * @returns Parsed TableCellProperties
   */
  private async parseTableCellPropertiesElement(tcPrElement: Record<string, unknown>): Promise<TableCellProperties> {
    const props: Partial<TableCellProperties> = {};

    // Parse table cell width (w:tcW)
    props.tcW = await this.parseTableCellWidth(tcPrElement);
    
    // Parse table cell borders (w:tcBorders)
    props.tcBorders = await this.parseTableCellBorders(tcPrElement);
    
    // Parse shading (w:shd)
    props.shd = await this.parseShading(tcPrElement);
    
    // Parse table cell margins (w:tcMar)
    props.tcMar = await this.parseTableCellMargins(tcPrElement);
    
    // Parse text direction (w:textDirection)
    props.textDirection = this.parseTextDirection(tcPrElement);
    
    // Parse vertical alignment (w:vAlign)
    props.vAlign = this.parseVerticalAlignment(tcPrElement);
    
    // Parse hide mark (w:hideMark)
    props.hideMark = this.parseHideMark(tcPrElement);
    
    // Parse cell merge (w:cellMerge)
    props.cellMerge = this.parseCellMerge(tcPrElement);
    
    // Parse grid span (w:gridSpan)
    props.gridSpan = this.parseGridSpan(tcPrElement);

    // Set convenience properties to match test expectations
    if (props.tcW?.width) {
      props.width = props.tcW.width;
    }
    if (props.tcBorders) {
      props.borders = props.tcBorders;
    }
    if (props.tcMar) {
      props.margins = props.tcMar;
    }
    if (props.shd?.fill) {
      props.shading = props.shd.fill;
    }

    return TableCellPropertiesModel.create(props);
  }

  /**
   * Parse table cell width from w:tcW element
   * @param tcPrElement - w:tcPr element
   * @returns TableWidth or undefined
   */
  private async parseTableCellWidth(tcPrElement: Record<string, unknown>): Promise<TableWidth | undefined> {
    const tcW = getFirstChildElement(tcPrElement, 'w:tcW');
    if (!tcW) return undefined;

    const type = this.getAttribute(tcW, 'w:type');
    const widthValue = this.getAttribute(tcW, 'w:w');
    
    let width: number | undefined;
    if (widthValue) {
      const parsedWidth = parseInt(widthValue, 10);
      if (!isNaN(parsedWidth)) {
        // Convert from twips to points if type is dxa
        width = type === 'dxa' ? convertTwipsToPoints(parsedWidth) : parsedWidth;
      }
    }

    return TableWidthModel.create({
      type,
      width,
    });
  }

  /**
   * Parse table cell borders from w:tcBorders element
   * @param tcPrElement - w:tcPr element
   * @returns TableCellBorders or undefined
   */
  private async parseTableCellBorders(tcPrElement: Record<string, unknown>): Promise<TableCellBorders | undefined> {
    const tcBorders = getFirstChildElement(tcPrElement, 'w:tcBorders');
    if (!tcBorders) return undefined;

    const borders: Partial<TableCellBorders> = {};

    // Parse each border direction
    for (const direction of ['top', 'left', 'bottom', 'right', 'insideH', 'insideV']) {
      const borderElement = getFirstChildElement(tcBorders, `w:${direction}`);
      if (borderElement) {
        const color = this.getAttribute(borderElement, 'w:color');
        const sizeValue = this.getAttribute(borderElement, 'w:sz');
        const spaceValue = this.getAttribute(borderElement, 'w:space');
        const val = this.getAttribute(borderElement, 'w:val');

        const size = sizeValue ? parseInt(sizeValue, 10) : undefined;
        const space = spaceValue ? parseInt(spaceValue, 10) : undefined;

        borders[direction as keyof TableCellBorders] = BorderPropertiesModel.create({
          color: color && color !== 'auto' ? `#${color}` : color,
          size: !isNaN(size as number) ? size : undefined,
          space: !isNaN(space as number) ? space : undefined,
          val,
        });
      }
    }

    return Object.keys(borders).length > 0 ? TableCellBordersModel.create(borders) : undefined;
  }

  /**
   * Parse shading from w:shd element
   * @param tcPrElement - w:tcPr element
   * @returns ShadingProperties or undefined
   */
  private async parseShading(tcPrElement: Record<string, unknown>): Promise<ShadingProperties | undefined> {
    const shd = getFirstChildElement(tcPrElement, 'w:shd');
    if (!shd) return undefined;

    const fill = this.getAttribute(shd, 'w:fill');
    const val = this.getAttribute(shd, 'w:val');
    const color = this.getAttribute(shd, 'w:color');

    return ShadingPropertiesModel.create({
      fill: fill && fill !== 'auto' ? `#${fill}` : fill,
      val,
      color: color && color !== 'auto' ? `#${color}` : color,
    });
  }

  /**
   * Parse table cell margins from w:tcMar element
   * @param tcPrElement - w:tcPr element
   * @returns MarginProperties or undefined
   */
  private async parseTableCellMargins(tcPrElement: Record<string, unknown>): Promise<MarginProperties | undefined> {
    const tcMar = getFirstChildElement(tcPrElement, 'w:tcMar');
    if (!tcMar) return undefined;

    const margins: Partial<MarginProperties> = {};

    // Parse each margin direction
    for (const direction of ['top', 'left', 'bottom', 'right']) {
      const marginElement = getFirstChildElement(tcMar, `w:${direction}`);
      if (marginElement) {
        const widthValue = this.getAttribute(marginElement, 'w:w');
        if (widthValue) {
          const parsedWidth = parseInt(widthValue, 10);
          if (!isNaN(parsedWidth)) {
            margins[direction as keyof MarginProperties] = convertTwipsToPoints(parsedWidth);
          }
        }
      }
    }

    return Object.keys(margins).length > 0 ? MarginPropertiesModel.create(margins) : undefined;
  }

  /**
   * Parse text direction from w:textDirection element
   * @param tcPrElement - w:tcPr element
   * @returns Text direction or undefined
   */
  private parseTextDirection(tcPrElement: Record<string, unknown>): string | undefined {
    const textDirection = getFirstChildElement(tcPrElement, 'w:textDirection');
    return textDirection ? this.getAttribute(textDirection, 'w:val') : undefined;
  }

  /**
   * Parse vertical alignment from w:vAlign element
   * @param tcPrElement - w:tcPr element
   * @returns Vertical alignment or undefined
   */
  private parseVerticalAlignment(tcPrElement: Record<string, unknown>): string | undefined {
    const vAlign = getFirstChildElement(tcPrElement, 'w:vAlign');
    return vAlign ? this.getAttribute(vAlign, 'w:val') : undefined;
  }

  /**
   * Parse hide mark from w:hideMark element
   * @param tcPrElement - w:tcPr element
   * @returns True if hide mark is present, false otherwise
   */
  private parseHideMark(tcPrElement: Record<string, unknown>): boolean {
    return extractBooleanProperty(tcPrElement, 'w:hideMark');
  }

  /**
   * Parse cell merge from w:cellMerge element
   * @param tcPrElement - w:tcPr element
   * @returns Cell merge value or undefined
   */
  private parseCellMerge(tcPrElement: Record<string, unknown>): string | undefined {
    const cellMerge = getFirstChildElement(tcPrElement, 'w:cellMerge');
    return cellMerge ? this.getAttribute(cellMerge, 'w:val') : undefined;
  }

  /**
   * Parse grid span from w:gridSpan element
   * @param tcPrElement - w:tcPr element
   * @returns Grid span value or undefined
   */
  private parseGridSpan(tcPrElement: Record<string, unknown>): number | undefined {
    const gridSpan = getFirstChildElement(tcPrElement, 'w:gridSpan');
    if (!gridSpan) return undefined;

    const val = this.getAttribute(gridSpan, 'w:val');
    if (val) {
      const parsed = parseInt(val, 10);
      return !isNaN(parsed) ? parsed : undefined;
    }
    return undefined;
  }
} 

File: docx_parser_converter_python/parsers/document/table_cell_properties_parser.py
"""Parser for table cell properties elements."""

from lxml.etree import _Element as Element

from models.document.table_cell import TableCellMargins, TableCellProperties
from parsers.common.border_parser import parse_table_borders
from parsers.common.shading_parser import parse_shading
from parsers.common.width_parser import parse_width
from parsers.utils import find_child, get_attribute, get_int_attribute, parse_toggle


def parse_table_cell_margins(element: Element | None) -> TableCellMargins | None:
    """Parse <w:tcMar> element.

    XML Example:
        <w:tcMar>
            <w:top w:w="72" w:type="dxa"/>
            <w:left w:w="115" w:type="dxa"/>
            <w:bottom w:w="72" w:type="dxa"/>
            <w:right w:w="115" w:type="dxa"/>
        </w:tcMar>

    Args:
        element: The <w:tcMar> element or None

    Returns:
        TableCellMargins model or None if element is None
    """
    if element is None:
        return None

    return TableCellMargins(
        top=parse_width(find_child(element, "top")),
        left=parse_width(find_child(element, "left")),
        bottom=parse_width(find_child(element, "bottom")),
        right=parse_width(find_child(element, "right")),
    )


def parse_table_cell_properties(element: Element | None) -> TableCellProperties | None:
    """Parse <w:tcPr> element.

    XML Example:
        <w:tcPr>
            <w:tcW w:w="2880" w:type="dxa"/>
            <w:gridSpan w:val="2"/>
            <w:vMerge w:val="restart"/>
            <w:shd w:val="clear" w:fill="FFFF00"/>
        </w:tcPr>

    Args:
        element: The <w:tcPr> element or None

    Returns:
        TableCellProperties model or None if element is None
    """
    if element is None:
        return None

    # Cell width
    tc_w = parse_width(find_child(element, "tcW"))

    # Cell borders
    tc_borders = parse_table_borders(find_child(element, "tcBorders"))

    # Shading
    shd = parse_shading(find_child(element, "shd"))

    # Cell margins
    tc_mar = parse_table_cell_margins(find_child(element, "tcMar"))

    # Text direction
    text_direction_elem = find_child(element, "textDirection")
    text_direction = (
        get_attribute(text_direction_elem, "val") if text_direction_elem is not None else None
    )

    # Vertical alignment
    v_align_elem = find_child(element, "vAlign")
    v_align = get_attribute(v_align_elem, "val") if v_align_elem is not None else None

    # Grid span (horizontal merge)
    grid_span_elem = find_child(element, "gridSpan")
    grid_span = get_int_attribute(grid_span_elem, "val") if grid_span_elem is not None else None

    # Vertical merge
    v_merge_elem = find_child(element, "vMerge")
    v_merge = None
    if v_merge_elem is not None:
        # If present without val, it means "continue"
        v_merge = get_attribute(v_merge_elem, "val") or "continue"

    # Horizontal merge (deprecated)
    h_merge_elem = find_child(element, "hMerge")
    h_merge = get_attribute(h_merge_elem, "val") if h_merge_elem is not None else None

    # No wrap
    no_wrap = parse_toggle(find_child(element, "noWrap"))

    # Fit text
    tc_fit_text = parse_toggle(find_child(element, "tcFitText"))

    # Hide mark
    hide_mark = parse_toggle(find_child(element, "hideMark"))

    return TableCellProperties(
        tc_w=tc_w,
        tc_borders=tc_borders,
        shd=shd,
        tc_mar=tc_mar,
        text_direction=text_direction,
        v_align=v_align,
        grid_span=grid_span,
        v_merge=v_merge,
        h_merge=h_merge,
        no_wrap=no_wrap,
        tc_fit_text=tc_fit_text,
        hide_mark=hide_mark,
    )
